backend/src/git.rs

```diff
@@ -332,65 +408,233 @@ impl Interface {
+    /// it will return a `color_eyre::eyre::Result`, which contains context about the failure.
+    fn set_upstream_if_exists(
+        &self,
+        repo: &git2::Repository,
+        branch_name: &str,
+        remote_name: &str,
+    ) -> Result<()> {
+        let remote_ref = format!("refs/remotes/{}/{}", remote_name, branch_name);
+
+        // Check if the remote branch exists
+        let remote_branch = repo.find_reference(&remote_ref)
+            .context(format!("Remote branch '{}' not found", remote_ref))?;
+
+        // Get the shorthand branch name
+        let remote_branch_name = remote_branch.shorthand()
+            .ok_or_else(|| color_eyre::eyre::eyre!("Failed to get shorthand name for remote branch '{}'", remote_ref))?;
+
+        info!(
+            "Setting upstream for local branch '{}' to remote '{}'",
+            branch_name, remote_branch_name
+        );
+
+        let mut branch = repo.find_branch(branch_name, git2::BranchType::Local)
+            .context("Failed to find local branch")?;
+
+        branch.set_upstream(Some(remote_branch_name))
+            .context("Failed to set upstream")?;
+
+        info!("Successfully set upstream for branch '{}' to '{}'", branch_name, remote_branch_name);
+
+        Ok(())
+    }
+
+    /// A code level re-implementation of `git fetch`.
+    /// `git fetch` will sync your local `origin/[BRANCH]` with the remote, but it won't
+    /// merge those changes into your local branch.
+    /// 
+    /// This implementation can fetch either all branches or a specified branch.
+    ///
+    /// Returns a reference to the latest commit fetched from remote (`FETCH_HEAD`).
+    fn git_fetch<'a>(repo: &'a Repository, branch: Option<&'a str>) -> Result<AnnotatedCommit<'a>> {
+        let mut remote = repo.find_remote("origin")?;
+
         let mut fetch_options = FetchOptions::new();
         fetch_options.download_tags(git2::AutotagOption::All);
-        // https://git-scm.com/book/en/v2/Git-Internals-The-Refspec
-        remote.fetch(
-            &["+refs/heads/*:refs/remotes/origin/*"],
-            Some(&mut fetch_options),
-            None,
-        )?;
+        
+        match branch {
+            Some(branch_name) => {
+                // Fetch only the specified branch
+                remote.fetch(
+                    &[&format!("refs/heads/{branch_name}:refs/remotes/origin/{branch_name}")],
+                    Some(&mut fetch_options),
+                    None,
+                )?;
+            },
+            None => {
+                // Fetch all branches
+                remote.fetch(
+                    &["+refs/heads/*:refs/remotes/origin/*"],
+                    Some(&mut fetch_options),
+                    None,
+                )?;
+            },
+        }
         drop(remote);
 
         let fetch_head = repo.find_reference("FETCH_HEAD")?;
+        let fetch_head_name = fetch_head.name().unwrap_or("unknown");
```

Comment:
I don't understand git well enough to understand this line of code, is there any reason you prefer setting it to `"unknown"` here, rather than propagating an error with `?`?